### PR TITLE
[D3D-2585] Adds debug visualisation shadow

### PR DESCRIPTION
**Pull request by Kuranes**
Show both target bbox as computed, and light shadow frustum as computed.

### CODE REVIEW

File: sources/osg/Shape.js [hunk 1]
```diff
@@ -730,6 +730,145 @@ var createGridGeometry = function ( cx, cy, cz, wx, wy, wz, hx, hy, hz, res1, re
...
+
+
+    g.getPrimitives().push( new DrawElements( PrimitiveSet.LINES, new BufferArray( 'ELEMENT_ARRAY_BUFFER', indexes, 1 ) ) );
+
+    return g;
+};
+// 
+
+/*
+ * allow for bounding boxes dynamically updated client side
```
Comment (cedricpinson) [inline sources/osg/Shape.js]: Rename and move updateBoundingBoxGeometryWithPerspectiveFrustum to osgUtil/debug.js

Could be named this way:

createFrustrumDebugGeometry: function( projection, dr ) {...}
You can create the debug geometry in the function and create a updateCallback that will update the geometry from the projection matrix

Comment (Kuranes) [inline sources/osg/Shape.js]: osgUtils/DebugBoundingBox.js ?


 "createFrustrumDebugGeometry"  :  frustum is a boundinbox geometry just "updated" ?
I use  createBoundingBoxGeometry for BoundinBox debug AND frustum debug.

More correct term would be "createLineBoxGeometry" 
that can be updated as a boundingbox or as a frustum ?

Should move all the additions  in "osgUtils/DebugBoundingBox.js" including  the create ?
or just the 2 updates ?

**Comment by Kuranes** [comment]
Heeded the comment :

- moved everything in osgUtils/Debug.js 
- "createBoundingBoxGeometry" named 'createBoxDebugLineGeometry'
- added createFrustrumDebugGeometry: function( projection, dr ) {...}


( the ability to "autoupdate" doesn't work in shadowmap though because matrix are reused/zero'ed elsewhere  )

**Comment by cedricpinson** [comment]
it's osgUtils/debug.js not Debug, upper case first letter is for class.

Because there is createAxisGeometry, createGridGeometry  in shape it's better to have createBoundingBoxGeometry in shape to be consistent.

I dont want to expose updateBoundingBoxGeometry and updateBoundingBoxGeometryWithPerspectiveFrustum in debug.js . If you need to update the debug element without reference, then add an update methodGeometry function:

var updateFrustrum =  function( args ) {
do your stuff
};

var createDebugFrustrumGeometry = function() {
var geom = shape.createBoundingBoxGeometry();
geom.updateGeometry = updateFrustrum;
}

could be a simple function assigned to the debug geometry or a prototype switch. Probably we can start only with a function assigned to the geometry.

Also the naming is not consistent
createBoxDebugLineGeometry is not consistent with createFrustrumDebugGeometry.
could simply:
- createDebugFrustrumGeometry

**Comment by Kuranes** [comment]
ok, will do.

Do I rename Shape.js to shape.js then ?
(maybe I could do a pass on all osgjs file to check consistency)

**Comment by jtorresfabra** [comment]
Hey @Kuranes 

> Do I rename Shape.js to shape.js then ?
> (maybe I could do a pass on all osgjs file to check consistency)

That could break a lot of user code. If you plan to do that please send an advice to the list.

**Comment by Kuranes** [comment]
will rename all "functions files" to lowercase then

**Comment by Kuranes** [comment]
- rebased
- made the Renaming

(tested everything; found things broken but not by this PR (tutorial arealight and loadmodel.))

File: sources/osgShadow/ShadowMap.js [hunk 2]
```diff
@@ -1035,6 +1083,37 @@ ShadowMap.prototype = MACROUTILS.objectLibraryClass( MACROUTILS.objectInherit( S
         //this._shadowReceiveAttribute = undefined;
         this._texture = undefined;
         this._shadowedScene = undefined;
+    },
+
+    setDebug: function ( a ) {
```
Comment (cedricpinson) [inline sources/osgShadow/ShadowMap.js]: fix the argument name. I guess a is a boolean ?

File: sources/osgUtil/debug.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+};
+/*
+ * Given a Camera, create a wireframe representation
+ *  of its view frustum
+ */
+var createDebugFrustrumGeometry = function () {
+
+    var g = Shape.createBoundingBoxGeometry();
+    g.updateGeometry = updateFrustum.bind( g );
```
Comment (cedricpinson) [inline sources/osgUtil/debug.js]: go for me

File: sources/osgUtil/osgUtil.js [hunk 4]
```diff
@@ -20,6 +21,7 @@ var WebVR = require( 'osgUtil/WebVR' );
 var osgUtil = {};
 
 osgUtil.Composer = Composer;
+osgUtil.Debug = Debug;
```
Comment (cedricpinson) [inline sources/osgUtil/osgUtil.js]: not fixed

**Comment by Kuranes** [comment]
- fixed renaming and osgUtil mixin

Shouldn't https://github.com/cedricpinson/osgjs/blob/master/sources/osgUtil/GizmoGeometry.js expose too it's internal debug geoms ? ( plan, line, quad, circle geometrY )

**Comment by Kuranes** [comment]
- added near/far fix for spotlight shadow bugs

File: sources/osgUtil/osgUtil.js [hunk 5]
```diff
@@ -1,10 +1,12 @@
 'use strict';
 var Composer = require( 'osgUtil/Composer' );
+var Debug = require( 'osgUtil/debug' );
```
Comment (cedricpinson) [inline sources/osgUtil/osgUtil.js]: should be var debug =  and not Debug.
First upper letter is for class in the code too

**Comment by Kuranes** [comment]
- renamed Debug to debug in js code.

**Pull request merged**